lib/ddtrace/configuration.rb

```diff
@@ -13,18 +18,50 @@ def configuration
     def configure(target = configuration, opts = {})
       if target.is_a?(Settings)
         yield(target) if block_given?
+
+        # Rebuild immutable components from settings
+        rebuild_components!(target)
+
+        target
       else
         PinSetup.new(target, opts).call
       end
     end
 
-    # Helper methods
-    def tracer
-      configuration.tracer
+    def_delegators \
+      :components,
+      :health_metrics,
+      :runtime_metrics,
+      :tracer
+
+    protected
+
+    def components
+      @components ||= Components.new(configuration)
     end
 
-    def runtime_metrics
-      tracer.writer.runtime_metrics
+    def rebuild_components!(configuration)
+      # Build new components
+      new_components = Components.new(configuration)
+
+      # Teardown old components if they exist
+      teardown_components!(@components, new_components) if instance_variable_defined?(:@components)
+
+      # Activate new components
+      @components = new_components
+    end
+
+    def teardown_components!(old, current)
+      # Shutdown the old tracer, unless it's still being used.
+      # (e.g. a custom tracer instance passed in.)
+      old.tracer.shutdown! unless old.tracer == current.tracer
+
+      # Shutdown the old metrics, unless they are still being used.
+      # (e.g. custom Statsd instances.)
+      old_statsd = [old.runtime_metrics.statsd, old.health_metrics.statsd].uniq
```

Comment:
Can `runtime_metrics` and `health_metrics` ever be the same object under any reasonable scenario?

Reply:
I believe so, I was thinking of this scenario:

```
custom_statsd = Datadog.Statsd.new(hostname: 'my-statsd-host')

Datadog.configure do |c|
  c.runtime_metrics.enabled = true
  c.runtime_metrics.statsd = custom_statsd
  c.diagnostics.health_metrics.enabled = true
  c.diagnostics.health_metrics.statsd = custom_statsd
end
```

Now I can't think of why someone would then follow this up by reconfiguring either metrics with a different instance of Statsd, but its probably a good idea to close them down if they are no longer in use (for memory), and checking if Statsd changed for each is insufficient (causes one to be closed prematurely.)